In [1]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [2]:
import os
os.environ['KAGGLE_USERNAME'] = 'BDA'
os.environ['KAGGLE_KEY'] = 'KGAT_3f3ceca6c3fc64f2daf6a7f63a5fd0ed'


In [3]:
!pip install kaggle

In [4]:
##CHANGE PATH ACCORDING THE NOTEBOOK LOCATION
os.chdir('/content/drive/MyDrive/BigData/Prima_Esercitazione') ##!!change path

In [6]:
!kaggle datasets download -d fedesoriano/body-fat-prediction-dataset
##CHANGE PATH ACCORDING THE NOTEBOOK LOCATION
!unzip /content/drive/MyDrive/BigData/Prima_Esercitazione/body-fat-prediction-dataset.zip ##!!change path

Dataset URL: https://www.kaggle.com/datasets/fedesoriano/body-fat-prediction-dataset
License(s): copyright-authors
body-fat-prediction-dataset.zip: Skipping, found more recently modified local copy (use --force to force download)
Archive:  /content/drive/MyDrive/BigData/Prima_Esercitazione/body-fat-prediction-dataset.zip
  inflating: bodyfat.csv             


In [8]:
import pandas as pd
df=pd.read_csv('/content/drive/MyDrive/BigData/Prima_Esercitazione/bodyfat.csv') ##!!change path
df.head(20)

,Density,BodyFat,Age,Weight,Height,Neck,Chest,Abdomen,Hip,Thigh,Knee,Ankle,Biceps,Forearm,Wrist
0,1.0708,12.3,23,154.25,67.75,36.2,93.1,85.2,94.5,59.0,37.3,21.9,32.0,27.4,17.1
1,1.0853,6.1,22,173.25,72.25,38.5,93.6,83.0,98.7,58.7,37.3,23.4,30.5,28.9,18.2
2,1.0414,25.3,22,154.00,66.25,34.0,95.8,87.9,99.2,59.6,38.9,24.0,28.8,25.2,16.6
3,1.0751,10.4,26,184.75,72.25,37.4,101.8,86.4,101.2,60.1,37.3,22.8,32.4,29.4,18.2
4,1.0340,28.7,24,184.25,71.25,34.4,97.3,100.0,101.9,63.2,42.2,24.0,32.2,27.7,17.7
5,1.0502,20.9,24,210.25,74.75,39.0,104.5,94.4,107.8,66.0,42.0,25.6,35.7,30.6,18.8
6,1.0549,19.2,26,181.00,69.75,36.4,105.1,90.7,100.3,58.4,38.3,22.9,31.9,27.8,17.7
7,1.0704,12.4,25,176.00,72.50,37.8,99.6,88.5,97.1,60.0,39.4,23.2,30.5,29.0,18.8
8,1.0900,4.1,25,191.00,74.00,38.1,100.9,82.5,99.9,62.9,38.3,23.8,35.9,31.1,18.2
9,1.0722,11.7,23,198.25,73.50,42.1,99.6,88.6,104.1,63.1,41.7,25.0,35.6,30.0,19.2


1. print dataset basic properties: number of cells, rows, column
2. check data types

In [19]:
print("Column types of dataframe: \n")
print(df.dtypes)
print("\n")
print("Columns of dataframe: \n")
print(df.columns)
print("\n")
print("Dimension of dataframe: \n")
print(df.shape)
print("\n")

Column types of dataframe: 

Density    float64
BodyFat    float64
Age          int64
Weight     float64
Height     float64
Neck       float64
Chest      float64
Abdomen    float64
Hip        float64
Thigh      float64
Knee       float64
Ankle      float64
Biceps     float64
Forearm    float64
Wrist      float64
dtype: object


Columns of dataframe: 

Index(['Density', 'BodyFat', 'Age', 'Weight', 'Height', 'Neck', 'Chest',
       'Abdomen', 'Hip', 'Thigh', 'Knee', 'Ankle', 'Biceps', 'Forearm',
       'Wrist'],
      dtype='object')


Dimension of dataframe: 

(252, 15)




3.  compute summary statistics ... is there something weird somewhere? (Check minimum Height)

In [24]:
print(df['Height'].min())
print(df['Height'].max())
print(df['Weight'].mean())
print(df['BodyFat'].mean())
print(df['BodyFat'].min())
print(df['BodyFat'].max())

29.5
77.75
178.92440476190475
19.150793650793652
0.0
47.5


Since the height is measured in inches, having a partecipant under 30 inches means having a 75cm tall person. This can be classified as *outlier*.
And it seems like that someone has 0.0 as BodyFat value. Preety strange, I will investigate and find out if it's a missing value or a NaN converted in 0.0 by the min function.

4. Is that a real outlier or just a typo? (filter patients to check)

In [30]:
df_fil = df[df['Height'] < 64]
print(df_fil)

    Density  BodyFat  Age  Weight  Height  Neck  Chest  Abdomen    Hip  Thigh  \
41    1.025     32.9   44   205.0    29.5  36.6  106.0    104.3  115.5   70.6   

    Knee  Ankle  Biceps  Forearm  Wrist  
41  42.5   23.7    33.6     28.7   17.4  


We can see that is only one guy under 64 inches height. Now we can assume that this was a **typo** and that those data were collected on average people.
This must be a **typo** for other reasons: this guy's weight is 205 lbs, that for a drawf would be impressivly high wieght. Since the bodyfat is almost 33% and we know that 29.5 is a typo, we deduce that the real height should be 69.5 inches (1.76m) and it has 30kg of fat mass and 60kg of lean mass, classyfing this example as **mildly obese**.

5. Compute the average Density of patients with BodyFat > 0.25

In [35]:
df_bf = df[df['BodyFat'] > 25]

print(df_bf['Density'].mean())

1.0319030303030305


6. Compute the average Density of patients with BodyFat < 0.2 Is this an intuitive result?


In [36]:
df_bf = df[df['BodyFat'] < 20]
print(df_bf['Density'].mean())

1.0704030303030303


Yes it is. Because we know that BodyFat measure the percentage of fat mass in our bodies, and the higher it gets, the lower the density will be.
This can be proved by the fact that muscle is way more dense than fat.

7. Add a BMI column: BMI = Weight[kg] / (Height[m])**2

In [45]:
def lbs_to_kg(x):
  return x*0.453

def inches_to_m(y):
  return y*0.0254

df['Weight'].map(lbs_to_kg)
df['Height'].map(inches_to_m)
df['BMI'] = df['Weight']/(df['Height']**2)
print(df['BMI'])
df.head(20)

0      0.033605
1      0.033189
2      0.035087
3      0.035392
4      0.036294
         ...   
247    0.029906
248    0.041315
249    0.042872
250    0.038378
251    0.042347
Name: BMI, Length: 252, dtype: float64


,Density,BodyFat,Age,Weight,Height,Neck,Chest,Abdomen,Hip,Thigh,Knee,Ankle,Biceps,Forearm,Wrist,BMI
0,1.0708,12.3,23,154.25,67.75,36.2,93.1,85.2,94.5,59.0,37.3,21.9,32.0,27.4,17.1,0.033605
1,1.0853,6.1,22,173.25,72.25,38.5,93.6,83.0,98.7,58.7,37.3,23.4,30.5,28.9,18.2,0.033189
2,1.0414,25.3,22,154.00,66.25,34.0,95.8,87.9,99.2,59.6,38.9,24.0,28.8,25.2,16.6,0.035087
3,1.0751,10.4,26,184.75,72.25,37.4,101.8,86.4,101.2,60.1,37.3,22.8,32.4,29.4,18.2,0.035392
4,1.0340,28.7,24,184.25,71.25,34.4,97.3,100.0,101.9,63.2,42.2,24.0,32.2,27.7,17.7,0.036294
5,1.0502,20.9,24,210.25,74.75,39.0,104.5,94.4,107.8,66.0,42.0,25.6,35.7,30.6,18.8,0.037628
6,1.0549,19.2,26,181.00,69.75,36.4,105.1,90.7,100.3,58.4,38.3,22.9,31.9,27.8,17.7,0.037204
7,1.0704,12.4,25,176.00,72.50,37.8,99.6,88.5,97.1,60.0,39.4,23.2,30.5,29.0,18.8,0.033484
8,1.0900,4.1,25,191.00,74.00,38.1,100.9,82.5,99.9,62.9,38.3,23.8,35.9,31.1,18.2,0.034879
9,1.0722,11.7,23,198.25,73.50,42.1,99.6,88.6,104.1,63.1,41.7,25.0,35.6,30.0,19.2,0.036698


8. define categorical BMI classes

In [47]:
df['cBMI']=pd.cut(df['BMI'],bins=[0,20,27,100],labels=['underweight','normal','overweight'])

8.1. groupby cBMI and compute average statistics

In [51]:
df_cBMI = df.groupby(['cBMI'])
df_cBMI.head(20)

/tmp/ipykernel_434/3788738311.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  df_cBMI = df.groupby(['cBMI'])


,Density,BodyFat,Age,Weight,Height,Neck,Chest,Abdomen,Hip,Thigh,Knee,Ankle,Biceps,Forearm,Wrist,BMI,cBMI
0,1.0708,12.3,23,154.25,67.75,36.2,93.1,85.2,94.5,59.0,37.3,21.9,32.0,27.4,17.1,0.033605,underweight
1,1.0853,6.1,22,173.25,72.25,38.5,93.6,83.0,98.7,58.7,37.3,23.4,30.5,28.9,18.2,0.033189,underweight
2,1.0414,25.3,22,154.00,66.25,34.0,95.8,87.9,99.2,59.6,38.9,24.0,28.8,25.2,16.6,0.035087,underweight
3,1.0751,10.4,26,184.75,72.25,37.4,101.8,86.4,101.2,60.1,37.3,22.8,32.4,29.4,18.2,0.035392,underweight
4,1.0340,28.7,24,184.25,71.25,34.4,97.3,100.0,101.9,63.2,42.2,24.0,32.2,27.7,17.7,0.036294,underweight
5,1.0502,20.9,24,210.25,74.75,39.0,104.5,94.4,107.8,66.0,42.0,25.6,35.7,30.6,18.8,0.037628,underweight
6,1.0549,19.2,26,181.00,69.75,36.4,105.1,90.7,100.3,58.4,38.3,22.9,31.9,27.8,17.7,0.037204,underweight
7,1.0704,12.4,25,176.00,72.50,37.8,99.6,88.5,97.1,60.0,39.4,23.2,30.5,29.0,18.8,0.033484,underweight
8,1.0900,4.1,25,191.00,74.00,38.1,100.9,82.5,99.9,62.9,38.3,23.8,35.9,31.1,18.2,0.034879,underweight
9,1.0722,11.7,23,198.25,73.50,42.1,99.6,88.6,104.1,63.1,41.7,25.0,35.6,30.0,19.2,0.036698,underweight
